rpcq/core-messages.py

```diff
@@ -44,37 +44,183 @@ class Resources(Message):
+    """
+
     iqs: List[float] = field(default_factory=list)
-    """The raw waveform envelope samples,
-          alternating I and Q values."""
+    """The raw waveform envelope samples, alternating I and Q values."""
+
+
+@dataclass(eq=False, repr=False)
+class TemplateWaveform(AbstractWaveform):
+    """
+    A waveform envelope defined for a specific frame. A templated waveform is defined by a parameterized pulseshape rather than explicit IQ values. The message specification does not enforce that the duration is implementable on the hardware.
+    """
+
+    duration: float
+    """Length of the pulse in seconds"""
+
+    scale: float = 1.e+0
+    """Scale to apply to waveform envelope"""
+
+    phase: float = 0.0e+0
+    """Phase [units of tau=2pi] to rotate the complex waveform envelope."""
+
+    detuning: float = 0.0e+0
+    """Modulation to apply to the waveform in Hz"""
+
+
+@dataclass(eq=False, repr=False)
+class GaussianWaveform(TemplateWaveform):
+    """
+    A Gaussian shaped waveform envelope defined for a specific frame.
+    """
+
+    fwhm: Optional[float] = None
+    """Full Width Half Max shape paramter in seconds"""
+
+    t0: Optional[float] = None
+    """Center time coordinate of the shape in seconds. Defaults to mid-point of pulse."""
+
+
+@dataclass(eq=False, repr=False)
+class DragGaussianWaveform(TemplateWaveform):
+    """
+    A Gaussian shaped waveform envelope defined for a specific frame.
+    """
+
+    fwhm: Optional[float] = None
+    """Full Width Half Max shape paramter in seconds"""
+
+    t0: Optional[float] = None
+    """Center time coordinate of the shape in seconds. Defaults to mid-point of pulse."""
+
+    anh: float = -2.1e+8
+    """Anharmonicity of the qubit, f01-f12 in (Hz)"""
+
+    alpha: float = 0.0e+0
+    """Dimensionless DRAG parameter"""
+
+
+@dataclass(eq=False, repr=False)
+class HermiteGaussianWaveform(TemplateWaveform):
+    """
+    Hermite-Gaussian shaped pulse. Reference: Effects of arbitrary laser ~
+      or NMR pulse shapes on population inversion and coherence Warren S. Warren. ~
+      81, (1984); doi: 10.1063/1.447644
+    """
+
+    fwhm: Optional[float] = None
+    """Full Width Half Max shape paramter in seconds"""
+
+    t0: float = None
+    """Center time coordinate of the shape in seconds. Defaults to mid-point of pulse."""
+
+    anh: float = -2.1e+8
+    """Anharmonicity of the qubit, f01-f12 in Hz"""
+
+    alpha: float = 0.0e+0
+    """Dimensionless DRAG parameter"""
+
+    second_order_hrm_coeff: float = 9.56e-1
+    """Second order coefficient (see paper)"""
+
+
+@dataclass(eq=False, repr=False)
+class ErfSquareWaveform(TemplateWaveform):
+    """
+    Pulse with a flat top and rounded shoulders given by error functions
+    """
+
+    risetime: float = 1.e-9
+    """The width of the rise and fall sections in seconds."""
+
+    pad_left: float = 0.0e+0
+    """Lentgh of zero-amplitude padding before the pulse."""
```

Comment:
Should add units to these.

Reply:
:roll_eyes: should probably also spell length correctly.